### PR TITLE
Feature/issue 2778 add add diag

### PR DESCRIPTION
**Pull request by drezap**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
Add the add_diag function to the language.

#### Intended Effect
Less for loops.

#### How to Verify
`make stan/src/test/test-models/good/function-signatures/math/matrix/add_diag`
with command stan.

#### Side Effects
None.

#### Documentation
I'll open up an issue to add to docs when this is merged.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
Copyright 2019 Andre Zapico

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/test/test-models/good/function-signatures/math/matrix/add_diag.stan [hunk 1]
```diff
@@ -0,0 +1,18 @@
+functions {}
+data {
+  int<lower=1> N;
+  matrix[N, N] K;
+  row_vector[N] row_vec;
+  vector[N] vec;
+  int scalar;
+}
+transformed data {
```
Comment (bob-carpenter) [inline src/test/test-models/good/function-signatures/math/matrix/add_diag.stan]: This file should have tests for all possible combinations of parameter and data for all possible signatures.   That's a total of 3 signatures * 4 instantiations each (data/data, data/parameter, parameter/data, parameter/parameter).

**Comment by bob-carpenter** [comment]
Can someone remind which is better, greater or less than one.   And also, what's the range of variation we expect.  

We don't want to induce 1% performance regressions on PRs, but how much of this is noise?

**Comment by wds15** [comment]
Numbers greater than 1 indicate a speed up...but smaller than 3% changes are probably noise anyway.

**Comment by drezap** [comment]
Is this ready?

**Pull request merged**